src/__private_api.rs

```diff
@@ -1,57 +1,116 @@
+    #[inline]
+    fn into_kvs(self) -> Option<&'a [(&'a str, &'a Value<'a>)]> {
+        Some(self)
+    }
+}
+
+impl<'a> KVs<'a> for () {
+    #[inline]
+    fn into_kvs(self) -> Option<&'a [(&'a str, &'a Value<'a>)]> {
+        None
+    }
+}
+
+// Log implementation.
+
+fn log_impl(
     args: Arguments,
     level: Level,
     &(target, module_path, file): &(&str, &'static str, &'static str),
     line: u32,
-    kvs: Option<&[(&str, &str)]>,
+    kvs: Option<&[(&str, &Value)]>,
 ) {
+    #[cfg(not(feature = "kv_unstable"))]
     if kvs.is_some() {
         panic!(
             "key-value support is experimental and must be enabled using the `kv_unstable` feature"
         )
     }
 
-    crate::logger().log(
-        &Record::builder()
-            .args(args)
-            .level(level)
-            .target(target)
-            .module_path_static(Some(module_path))
-            .file_static(Some(file))
-            .line(Some(line))
-            .build(),
-    );
+    let mut builder = Record::builder();
+
+    builder
+        .args(args)
+        .level(level)
+        .target(target)
+        .module_path_static(Some(module_path))
+        .file_static(Some(file))
+        .line(Some(line));
+
+    #[cfg(feature = "kv_unstable")]
+    builder.key_values(&kvs);
+
+    crate::logger().log(&builder.build());
 }
 
-#[cfg(feature = "kv_unstable")]
-pub fn log(
+pub fn log<'a, K>(
     args: Arguments,
     level: Level,
-    &(target, module_path, file): &(&str, &'static str, &'static str),
+    target_module_path_and_file: &(&str, &'static str, &'static str),
     line: u32,
-    kvs: Option<&[(&str, &dyn crate::kv::ToValue)]>,
-) {
-    crate::logger().log(
-        &Record::builder()
-            .args(args)
-            .level(level)
-            .target(target)
-            .module_path_static(Some(module_path))
-            .file_static(Some(file))
-            .line(Some(line))
-            .key_values(&kvs)
-            .build(),
-    );
+    kvs: K,
+) where
+    K: KVs<'a>,
+{
+    log_impl(
+        args,
+        level,
+        target_module_path_and_file,
+        line,
+        kvs.into_kvs(),
+    )
 }
 
 pub fn enabled(level: Level, target: &str) -> bool {
     crate::logger().enabled(&Metadata::builder().level(level).target(target).build())
 }
+
+/// This function is not intended to be actually called by anyone. The intention of this function is for forcing the
+/// compiler to generate monomorphizations of the `log` function, so that they can be shared by different downstream
+/// crates.
+///
+/// The idea is that with [`share-generics`](https://github.com/rust-lang/rust/pull/48779), downstream crates can reuse
+/// generic monomorphizations from upstream crates, but not siblings crates. So it is best to instantiate these
+/// monomorphizations in the `log` crate, so downstream crates are guaranteed to be able to share them.
+pub fn instantiate_log_function() -> usize {

```

Comment:
This seems a bit obtuse to include here. Based on your binary size profiling I'm just not sure this is pulling its weight. Since we don't have CI in place to monitor the impact of binary size it's possible the optimization it indirectly encourages could be undone at any point without our knowing.

So I think we should remove this function.

Reply:
OK, I have `removed instantiate_log_function`.